Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -31,8 +31,121 @@
+
+struct __CGPath {
+    CFRuntimeBase _base;
+
+    ComPtr<ID2D1PathGeometry> _pathGeometry;
+    ComPtr<ID2D1GeometrySink> _geometrySink;
+
+    bool _isPathClosed;
+    bool _isFigureClosed;
+    CGPoint _currentPoint;
+    CGPoint _startingPoint;
+
+    // A private helper function for re-opening a path geometry. CGPath does not have a concept of an open
+    // and a closed path but D2D relies on it. A path/sink cannot be read from while the path is open
+    // thus it must be closed. However, a CGPath can be edited again after being read from so we must open
+    // the path again. This cannot be done normally, so we must create a new path with the old path information
+    // to edit.
+    void __preparePathForEditing() {
+        if (_isPathClosed) {
+            _pathGeometry.~ComPtr();
+            _geometrySink.~ComPtr();
+
+            // Re-open this geometry.
+            ComPtr<ID2D1Factory> factory = _CreateD2DFactoryInstance();
+            if (FAILED(factory->CreatePathGeometry(&_pathGeometry))) {
+                FAIL_FAST();
+            }
+            if (FAILED(_pathGeometry->Open(&_geometrySink))) {
+                FAIL_FAST();
+            }
+
+            // This path is no longer closed.
+            _isPathClosed = false;
+            // Without a new figure being created, it's by default closed
+            _isFigureClosed = true;
+        }
+    }
+
+    void __closePath() {
+        if (!_isPathClosed) {
+            __endFigure(D2D1_FIGURE_END_OPEN);
+            _geometrySink->Close();
+            _isPathClosed = true;
+        }
+    }
+
+    void __beginFigure() {
+        if (_isFigureClosed) {
+            _geometrySink->BeginFigure(D2D1::Point2F(_currentPoint.x, _currentPoint.y), D2D1_FIGURE_BEGIN_FILLED);
+            _isFigureClosed = false;
+        }
+    }
+
+    void __endFigure(D2D1_FIGURE_END figureStatus) {
+        if (!_isFigureClosed) {
+            _geometrySink->EndFigure(figureStatus);
+            _isFigureClosed = true;
+        }
+    }
+};
+
+static CFBooleanRef __CGPathEqual(CFTypeRef cf1, CFTypeRef cf2) {
+    __CGPath* path1 = (__CGPath*)cf1;
+    __CGPath* path2 = (__CGPath*)cf2;
+
+    return kCFBooleanFalse;
+}
+
+static void __CGPathInit(CFTypeRef cf) {
+    CGPathRef path = (CGPathRef)cf;
+    // ComPtr needs to be manually new'd, to avoid calling Create functions to an incomplete out-pointer
+    __CGPath* mutablePath = const_cast<__CGPath*>(path);
+    new (std::addressof(mutablePath->_pathGeometry)) ComPtr<ID2D1PathGeometry>();
+    new (std::addressof(mutablePath->_geometrySink)) ComPtr<ID2D1GeometrySink>();
+
+    mutablePath->_currentPoint = { 0, 0 };
+    mutablePath->_startingPoint = { 0, 0 };
+    mutablePath->_isFigureClosed = true;
+}
+
+static void __CGPathDeallocate(CFTypeRef cf) {
+    CGPathRef path = (CGPathRef)cf;
+    // ComPtr needs to be manually destructed, since CFTypes do a memset(0) on dealloc
+    path->_pathGeometry.~ComPtr();
```

Comment:
>ComPtr [](start = 25, length = 6)

ComPtr doesnt allocate any memory on its own.  You should assign nullptr or Release() instead.